src/gluonts/mx/trainer/model_iteration_averaging.py

```diff
@@ -0,0 +1,286 @@
+            # load the cached model
+            for name, param_cached in self.cached_model.items():
+                model.collect_params()[name].set_data(param_cached)
+
+
+class NTA(IterationAveragingStrategy):
+    r"""
+    Implement Non-monotonically Triggered AvSGD (NTA).
+    This method is based on paper "Regularizing and Optimizing LSTM Language Models", 
+    (https://openreview.net/pdf?id=SyyGPP0TZ), and an implementation is available in Salesforce GitHub 
+    (https://github.com/salesforce/awd-lstm-lm/blob/master/main.py)
+    Note that it mismatches the arxiv (and gluonnlp) version, which is referred to as NTA_V2 below
+    """
+
+    val_logs: List[Any]
+
+    @validated()
+    def __init__(
+        self,
+        n: int = 5,
+        maximize: bool = False,
+        last_n_trigger: bool = False,
+        eta: float = 0,
+    ):
+        r"""
+        Depending on the choice of metrics, the users may want to minimize or maximize the metrics.
+        Thus, set maximize = True to maximize, otherwise minimize.
+
+        Parameters
+        ----------
+        n
+            The non-montone interval.
+        maximize
+            Whether to maximize or minimize the validation metric.
+        eta
+            Parameter of polynomial-decay averaging.
+        last_n_trigger
+            If True, use [-n:] in average trigger, otherwise use [:-n]
+        """
+
+        super().__init__(eta=eta)
+
+        self.n = n
+        self.maximize = maximize
+        self.last_n_trigger = last_n_trigger
+        # Historical validation metrics.
+        self.val_logs = []
+
+    def update_average_trigger(
+        self, metric: Any = None, epoch: int = 0, **kwargs
+    ):
+        r"""
+        Parameters
+        ----------
+        metric
+            The criteria to trigger averaging.
+        epoch
+            The epoch to start averaging, not used in NTA
+
+        Returns
+        -------
+        """
+
+        if not self.averaging_started and self.n > 0:
+            if self.last_n_trigger:
+                if self.maximize:
+                    if len(self.val_logs) >= self.n and metric < max(
+                        self.val_logs[-self.n :]
+                    ):
+                        self.averaging_started = True
+                else:
+                    if len(self.val_logs) >= self.n and metric > min(
+                        self.val_logs[-self.n :]
+                    ):
+                        self.averaging_started = True
+            else:
+                if self.maximize:
+                    if len(self.val_logs) > self.n and metric < max(
+                        self.val_logs[: -self.n]
+                    ):
+                        self.averaging_started = True
+                else:
+                    if len(self.val_logs) > self.n and metric > min(
+                        self.val_logs[: -self.n]
+                    ):
+                        self.averaging_started = True
```

Comment:
I guess this could be reformatted to something like:
```
n_val_logs = self.val_logs[-self.n :] if self.last_n_trigger else self.val_logs[: -self.n]
if self.maximize:
    if len(self.val_logs) >= self.n and metric < max(
        n_val_logs
    ):
        self.averaging_started = True
else:
    if len(self.val_logs) >= self.n and metric > min(
        n_val_logs
    ):
        self.averaging_started = True

```
in order to avoid all the repetition.

Reply:
For V2, it uses the last n elements in val_logs, so we only need to check "len(self.val_logs) >= self.n".
However, for V1, it uses the elements except the last n, so there must be at least self.n+1 elements in val_logs, so we need to check "len(self.val_logs) > self.n" instead.
I could still merge these 2 cases anyway, but maybe the current one is clearer to show the difference between them?

Reply:
Sure... feel free to ignore this comment if you think the existing format is more clear since this is more important.